modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala

```diff
@@ -155,16 +153,25 @@ object GlobalSnapshotConsensusFunctions {
       } yield (globalSnapshot, returnedEvents)
     }
 
-    private def updateBalancesByTransactions(balances: SortedMap[Address, Balance], txs: SortedSet[RewardTransaction]) =
-      txs.foldLeft(balances) {
-        case (acc, tx) =>
-          acc.updatedWith(tx.destination) { existingBalance =>
-            existingBalance
-              .flatMap(_.plus(tx.amount).toOption)
-              .orElse(Some(Balance.fromAmount(tx.amount)))
-          }
+    private def updateBalancesByRewards(
+      balances: SortedMap[Address, Balance],
+      txs: SortedSet[RewardTransaction]
+    ): (SortedSet[RewardTransaction], SortedMap[Address, Balance]) = {
+      val (toBurn, toUpdate) = txs.toList.map { tx =>
+        balances
+          .getOrElse(tx.destination, Balance.empty)
+          .plus(tx.amount)
+          .map((tx.destination, _))
+          .leftMap(_ => tx)
+      }.partitionBifold(identity)
+
+      val updatedBalances = toUpdate.foldLeft(balances) {
+        case (acc, (address, balance)) => acc.updated(address, balance)
```

Comment:
This will work incorrectly if there are 2 reward txs for some address.  Please replace it with a version that accumulates balances.

```scala
private def acceptRewardTxs(
  balances: SortedMap[Address, Balance],
  txs: SortedSet[RewardTransaction]
): (SortedMap[Address, Balance], SortedSet[RewardTransaction]) =
  txs.foldLeft((balances, SortedSet.empty[RewardTransaction])) { (acc, tx) =>
    val (updatedBalances, acceptedTxs) = acc

    updatedBalances
      .getOrElse(tx.destination, Balance.empty)
      .plus(tx.amount)
      .map(balance => (updatedBalances.updated(tx.destination, balance), acceptedTxs + tx))
      .getOrElse(acc)
  }
```

Reply:
Ooops, how could I not have seen it 🙈

Reply:
I see that you are collecting accepted transactions instead of failed ones but it is fine because I can just subtract accepted from all to get incorrect ones.